# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/EditorUtils.cpp
# Commit: 32d794b2a45a
# Full Hash: 32d794b2a45a300cd0f5b07dcac5ab5a8a89fb23
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-08-04 09:46:07
# Regressor Bug: 1774704
# File Overlap Count: 1
# Description:
#   Bug 1774704 - part 4-4: Make `HTMLEditor::ChangeSelectedHardLinesToList` stop setting `TopLevelEditSubActionData::mNewBlockElement` r=m_kato
#   
#   With taking an `AutoRangeArray` which is initialized with `Selection`, it can
#   restore selection and check if it's collapsed and if collapsed in the expected
#   list element.  Then, its caller can apply the returned range to `Selection`
# ==============================================================================

diff -r 23d50b5617de -r 32d794b2a45a editor/libeditor/EditorUtils.cpp
--- a/editor/libeditor/EditorUtils.cpp	Thu Aug 04 02:12:03 2022 +0000
+++ b/editor/libeditor/EditorUtils.cpp	Thu Aug 04 02:28:11 2022 +0000
@@ -85,6 +85,16 @@
   Initialize(aSelection);
 }
 
+AutoRangeArray::AutoRangeArray(const AutoRangeArray& aOther)
+    : mAnchorFocusRange(aOther.mAnchorFocusRange),
+      mDirection(aOther.mDirection) {
+  mRanges.SetCapacity(aOther.mRanges.Length());
+  for (const OwningNonNull<nsRange>& range : aOther.mRanges) {
+    RefPtr<nsRange> clonedRange = range->CloneRange();
+    mRanges.AppendElement(std::move(clonedRange));
+  }
+}
+
 template <typename PointType>
 AutoRangeArray::AutoRangeArray(const EditorDOMRangeBase<PointType>& aRange) {
   MOZ_ASSERT(aRange.IsPositionedAndValid());